### ISSUE TITLE
BatchNorm2d: Different memory format have different output

### ISSUE THREAD

**Issue by robert1003** [pytorch/pytorch#121651]
### 🐛 Describe the bug

Contiguous memory format model produce results that are not numerically similar to its channels_last memory format counterpart. More specifically, when `BatchNorm2d` is introduced, models with different memory format will have different parameters depending on the input memory format (contiguous_format as `cont`, channels_last as `last`:
* `cont` model + `last` data == `last` model + `last` data
* `cont` model + `cont` data != `last` model + `cont` data
* `cont` model + `cont` data != `last` model + `last` data
* `cont` model + `last` data == `last` model + `cont` data

Any ideas why is this happening? Thank you!

```python
##### IMPORTS
import os
os.environ["NCCL_ALGO"] = "Ring"
os.environ["NCCL_DEBUG"] = "INFO"

import torch
torch.set_printoptions(8)

import numpy as np
import random

##### UTILS

def set_deterministic(mode=True, warn_only=False):
    if mode:
        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
    elif "CUBLAS_WORKSPACE_CONFIG" in os.environ:
        del os.environ["CUBLAS_WORKSPACE_CONFIG"]
    torch.backends.cudnn.deterministic = mode
    torch.backends.cudnn.benchmark = not mode
    torch.use_deterministic_algorithms(mode, warn_only=warn_only)
    torch.backends.cuda.enable_mem_efficient_sdp(not mode)
    torch.backends.cuda.enable_flash_sdp(not mode)

def seed_all_rng(seed=10):
    # torch sets on both cpu and cuda
    torch.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)

def allclose(a, b):
    return torch.allclose(a.detach(), b.detach(), atol=1e-5, equal_nan=True)

##### SETUP

set_deterministic(warn_only=True)

batch_size, num_channels, h, w = 2, 1, 3, 3
num_iters = 10

def build_model():
	return torch.nn.Sequential(
            torch.nn.Conv2d(num_channels, num_channels, 3, 1, padding=1, dilation=1, groups=1, bias=True, padding_mode='zeros'),
            torch.nn.BatchNorm2d(num_channels),
            torch.nn.ReLU6(inplace=True),
        ).cuda()

def build_data():
    seed_all_rng(42)
    return torch.rand(batch_size, num_channels, h, w, requires_grad=False).cuda()

##### TEST

def test():
    data = build_data()

    seed_all_rng(42)
    model_channels_last = build_model()
    torch.nn.utils.convert_conv2d_weight_memory_format(model_channels_last, torch.channels_last)
    optimizer_channels_last = torch.optim.Adam(model_channels_last.parameters())

    seed_all_rng(42)
    model_contiguous = build_model()
    #model_contiguous.eval()
    #torch.nn.utils.convert_conv2d_weight_memory_format(model_contiguous, torch.channels_last)
    #model_contiguous.train()
    optimizer_contiguous = torch.optim.Adam(model_contiguous.parameters())

    for i in range(num_iters):
        seed_all_rng(i)

        assert allclose(model_contiguous[0].weight, model_channels_last[0].weight), \
            f"{model_contiguous[0].weight} \n---\n {model_channels_last[0].weight}"
        assert allclose(model_contiguous[0].bias, model_channels_last[0].bias), \
            f"{model_contiguous[0].bias} \n---\n {model_channels_last[0].bias}"
        assert allclose(model_contiguous[1].weight, model_channels_last[1].weight), \
            f"{model_contiguous[1].weight} \n---\n {model_channels_last[1].weight}"
        assert allclose(model_contiguous[1].bias, model_channels_last[1].bias), \
            f"{model_contiguous[1].bias} \n---\n {model_channels_last[1].bias}"

        # pass
        # baseline_result = model_contiguous(
        #     data.to(memory_format=torch.channels_last))
        # candidate_result = model_channels_last(
        #     data.to(memory_format=torch.channels_last))
        # fail
        baseline_result = model_contiguous(
            data.to(memory_format=torch.contiguous_format))
        candidate_result = model_channels_last(
            data.to(memory_format=torch.contiguous_format))
        # fail
        # baseline_result = model_contiguous(
        #     data.to(memory_format=torch.contiguous_format))
        # candidate_result = model_channels_last(
        #     data.to(memory_format=torch.channels_last))
        # pass
        # baseline_result = model_contiguous(
        #     data.to(memory_format=torch.channels_last))
        # candidate_result = model_channels_last(
        #     data.to(memory_format=torch.contiguous_format))

        assert allclose(baseline_result, candidate_result)

        baseline_result.sum().backward()
        candidate_result.sum().backward()

        optimizer_contiguous.step()
        optimizer_channels_last.step()

        optimizer_contiguous.zero_grad(set_to_none=1)
        optimizer_channels_last.zero_grad(set_to_none=1)
        
        print('iter', i)

test()
```

### Versions

PyTorch version: 2.1.0a0+git4c55dc5
Is debug build: False
CUDA used to build PyTorch: 12.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: 16.0.6 (https://github.com/conda-forge/clangdev-feedstock baacf2494cb86cecd0e9a29c840e78feb01a6ef3)
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-76-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 535.104.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] mypy-protobuf==3.3.0
[pip3] numpy==1.23.5
[pip3] pytorch-triton-rocm==2.2.0
[pip3] torch==2.1.1+1707334128.cuda122.cudnn894.ap
[pip3] torch-mlir==1695850709+71d8b6f2.1696652168
[pip3] torch-scatter==2.1.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchfile==0.1.0
[pip3] torchmetrics==1.3.1
[pip3] torchvision==0.16.0+fbb4cc5
[pip3] triton==2.1.0
[conda] magma-cuda121             2.6.1                         1    pytorch
[conda] mkl                       2024.0.0         ha957f24_49657    conda-forge
[conda] mkl-include               2024.0.0         ha957f24_49657    conda-forge
[conda] nomkl                     2.0                           0    anaconda
[conda] numpy                     1.23.5          py310h53a5b5f_0    conda-forge
[conda] pytorch-triton-rocm       2.2.0                    pypi_0    pypi
[conda] torch                     2.1.1+1707334128.cuda122.cudnn894.ap          pypi_0    pypi
[conda] torch-mlir                1695850709+71d8b6f2.1696652168          pypi_0    pypi
[conda] torch-scatter             2.1.0                    pypi_0    pypi
[conda] torch-tb-profiler         0.4.3                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchmetrics              1.3.1              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.16.0+fbb4cc5           pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by Chillee** [comment]
In general, there are no guarantees that kernels will have the same output for different memory layouts.

Different memory layouts may result in different kernels chosen, resulting in different reduction orders, resulting in value divergences.

**Comment by robert1003** [comment]
I see, thank you!
